### ISSUE TITLE
More sophisticated state representation for e.g. Policy Gradients

### ISSUE THREAD

**Issue by jeremysalwen** [deepmind/open_spiel#291]
As far as I can tell, in most of the cases where neural nets are used as part of a learning algorithm, the structure of the neural network is basically specified by the algorithm class, with only basic tunable knobs (e.g. [hidden_layer_sizes](https://github.com/deepmind/open_spiel/blob/754e0d7f32d87115f58b811e6b2db1954dfc080e/open_spiel/python/algorithms/policy_gradient.py#L105) for policy gradient algorithms).

The inputs and outputs are assumed to be fixed size unstructured vectors (i.e. every layer is just fully connected).  This makes sense for simple games like kuhn poker, but for more complex games, it can be important for the networks to be structured in a way that encodes the observations and action space in a way that is tailored to the structure of the game (for an extreme example, the [alphastar network](https://www.researchgate.net/figure/Overview-of-the-architecture-of-AlphaStar-A-detailed-description-is-provided-in-the_fig3_336911787)).

For a one-off project, I could imagine just ripping out the neural network code of the policy gradient algorithm, and replacing it with my own tensorflow code that only works for my specific game.  However, this does not seem like a scalable way to implement things.  Has there been any thought into making algorithms *parameterizeable* over the neural network structure?  

p.s. Another way of looking at it is that the observation_tensor destroys all structure in the observations, and the action representation as a discrete list destroys all structure of the action space.  Is there any way we could put some awareness of the action/observation space structure into Open Spiel?

**Comment by lanctot** [comment]
Hi @jeremysalwen ,

We have `ObservationShape` to specify the shape to reshape the observations into planar form for e.g. convolutional networks. Several of the perfect info games are laid out in planar form that allows one reshape on the python side (see Breakthrough, Tic-Tac-Toe).

The action space one has come up a number of times, and I even started writing it, e.g. a tensor-based action space. It is still on the todo list, but we have not really needed it yet.

But agreed support for complex observations and actions is limited. I am open to discuss how to do this but it is a huge change/addition, and would make the code much less simple. So there is a trade-off. We could define a new notion of a complex observation and games could return an object that agents could query.

**Comment by jeremysalwen** [comment]
Thanks @lanctot 

A few notes:
1. Adding more structure to the observations would not necessarily require changing the existing API, just adding additional data.  Just like the ObservationShape allows the algorithm to reshape the observation tensor before passing it to the NN, additional metadata could allow it to use the exact same observation tensor in a more principled way.  For example, I could imagine an API where the game could optionally annotate the observation tensor with information about its structure
``
TensorStructure([Spatial([16,16], Vector(100)), Vector(7)])
``
would tell us that the first 25600 elements of the tensor form a 16x16 grid of 100 dimensional vectors, and the last 7 elements form an unstructured vector.  Likely the first part would be fed through a convolutional layer, and the second part through fully connected layers.

2. However, even though it seems nice to try to separate the specification of the observation data structure from the network structure, I think the line is too ambiguous for the first approach to really work.  Another option is that we could define a new type of object -- a "Neural Encoder" which can be written for any game, that would define the "Torso" of the network for a given game.  The RL algorithm could still attach the heads it needs, but this would allow users to define multiple encodings/network structures to use for the same game, while easily swapping them out for each other, and reusing them across different RL algorithms.

I could see this being implemented on top of the existing observation tensor (more modular), or allowing the neural encoder to be more tightly coupled to the game implementation, i.e. it directly reads off the internal variables of the State implementation for the game.

3. Even with a more powerful observation encoding scheme, there is still a problem for my use case (implementing a card game) with the way actions are encoded.  I would like to be able to tie actions to specific cards, using something like Pointer Networks.  Writing an algorithm which is customized to my game, this is easy to do, but using a generic interface, I think it would take some novel design to get it to work properly.

This also suggests that we could take the "Neural Encoder" idea a step further, and let the entire network be swapped out, including the "heads" of the network.  This introduces some problems, because now the neural network definition will not just be tied to the game, but also the algorithm it is used for (This might not be so bad since many related algorithms require similarly structured NNs).

**Comment by michalsustr** [comment]
I like the first suggestion, i.e. add some TensorStructure. There can be a default neural encoder implementation (for the ML framework of choice) that takes the tensor structure as input and creates an "observation head" for the game.

**Comment by michalsustr** [comment]
It is worth thinking about also how to deal with the different kinds of observations. Right now, there is InformationState/Observation/PublicObservation/PrivateObservation. Each of them provides a string and a vector of doubles for tensor observations (except the last two ones because we don't have the use case yet). It just starts to feel quite crowded :)

**Comment by lanctot** [comment]
These are all great ideas. I agree it can be done without changing the existing core API. 

I think one thing that needs to be mentioned is the trade-offs between appealing to the masses and just doing what you need to do to get your own research done. OpenSpiel was designed in a way to make it easy to just tack-on what you need or change something you don't like without ever giving back, and that's ok. I'm particularly wary of big changes like this that add a fair amount of complexity if only a small fraction of the user base would benefit from it. 

So it's not clear to me yet that this is worth the price we'd pay in terms of added complexity, but I would entertain being convinced if the following points could be satisfied:

1. **Algorithm-independent**. We simply don't have the resources to maintain multiple copies of algorithms parameterized by the algorithm. And I don't see any way we could justify this added complexity.

2. **Easy C++ / Python inter-operability** Some of the proposed ideas might be difficult to wrap nicely using just pybind11. 

3. **Machine Learning Framework independent**. Sounds like the proposal would benefit from being done with a single framework in mind, but other than the algorithm implementations OpenSpiel is mostly ML framework independent, and people are free to use whatever they want (TF / JAX / PyTorch). This is an appealing property that I would not want to lose.

4. **A Dedicated Contributor** Probably most important of all: someone would have to implement this. We can discuss proposals all we want, but if there's nobody to write it, it won't happen. I don't think we have the bandwidth for such a large change... @jeremysalwen are you offering to write this (assuming we all agree on something)?

**Comment by jeremysalwen** [comment]
@lanctot I certainly am open to the idea of implementing the code.  I am not sure if I can commit to longer term maintenance, but I expect that either this feature will gain traction with other users, or if not, it could be removed easily.

1. **Algorithm-independent.** 100% on board here.
2. **Easy C++ / Python inter-operability** I am not so familiar with how python bindings interact with the various ML frameworks.  I would have to get a better understanding of how things work to know how easy it would be.
3.  **Machine Learning Framework independent**. I think the most valuable improvements would necessarily be dependent on the machine learning framework.  But i take this to mean more like *agnostic* to the ML framework?  i.e. new frameworks could be easily added, and no one is required or "preferred"?
4. **A Dedicated Contributor.** As I said, I might be able to perform this role.  However, I feel like I would want to discuss proposals more before committing (some sort of catch-22 happening here ;))

Specifically, I would want to make sure that what I was implementing would be powerful enough to support the project that I am using it for.  I have some idea (elaborated a bit above) about how to implement more sophisticated state representation in a very generic way.  

However, I see

1. The support for expressing relationships between the action space and observation space.
2. The support for auto-regressive policy selection.

as necessary pieces for which I do *not* have a clear idea of how to support.  (Note that I don't mean support needs to be baked into open_spiel, just that it would be possible for a user to implement these features on top of the API provided).  I think part of the solution may come from understanding open-spiel better, but also part of it may come from a nuanced discussion.

**Comment by lanctot** [comment]
@jeremysalwen you might be interested in the latest commits. @elkhrt has designed a new observation API that allows for richer representations. It does not touch everything you've brought up, but certainly some of the points about the observations. I can let Ed say more.

**Comment by elkhrt** [comment]
I've added structure to observations, currently only for Kuhn poker, leduc
poker, and coordinated matching pennies. But it's easy(ish) to add more.
Probably best to check out the update on Monday (I suggest looking at the
leduc playthroughs as a good starting point). Let me know what you think
and if you have any questions.

I haven't addressed structure in the action space at all. We could
statically return a set of action space partitions with associated
information perhaps? Maybe here you could write a straw man proposal and we
could discuss.

**Comment by jeremysalwen** [comment]
Thank you for keeping me updated.   I have been busy with a move, but I have not lost interest and I will take a look in a week or two!
